File: backend/ts_template/devdonalds.ts
import express, { Request, Response } from "express";
import HTTPError from 'http-errors';


// ==== Type Definitions, feel free to add or modify ==========================
interface cookbookEntry {
  name: string;
  type: string;
}

interface requiredItem {
  name: string;
  quantity: number;
}

interface recipe extends cookbookEntry {
  requiredItems: requiredItem[];
}

interface ingredient extends cookbookEntry {
  cookTime: number;
}

interface summary extends cookbookEntry {
  cookTime: number;
  requiredItems: requiredItem[];
}

// =============================================================================
// ==== HTTP Endpoint Stubs ====================================================
// =============================================================================
const app = express();
app.use(express.json());

// Store your recipes here!
const cookbook: any = [];

// Task 1 helper (don't touch)
app.post("/parse", (req:Request, res:Response) => {
  const { input } = req.body;

  const parsed_string = parse_handwriting(input)
  if (parsed_string == null) {
    res.status(400).send("this string is cooked");
    return;
  } 
  res.json({ msg: parsed_string });
  return;
  
});

// [TASK 1] ====================================================================
// Takes in a recipeName and returns it in a form that 
const parse_handwriting = (recipeName: string): string | null => {
  recipeName =  recipeName
    .replace(/[-_]/g, " ", )  // replaces - and _ to whitespace
    .toLowerCase()
    .replace(/[^a-z ]/g, "")  // removes all symbols
    .trim()
    .replace(/\s+/g, " ")     // remove duplicate whitespace
    .split(' ')
    .map((word) => word.charAt(0).toUpperCase() + word.substring(1)) // capitalise first letter
    .join(' ');

  if (!recipeName.length) {
    return null;
  }
  return recipeName;
}

// [TASK 2] ====================================================================
// Endpoint that adds a CookbookEntry to your magical cookbook
app.post("/entry", (req:Request, res:Response) => {
  const input: recipe | ingredient = req.body;

  res.status(200).send(add_entry(input));
});

// adds a new entry to cookbook
const add_entry = (input: recipe | ingredient) : void => {
  
  if (!(input.type === "recipe" || input.type === "ingredient")) {
    throw (HTTPError(400));
  } else if (cookbook.findIndex((entry) => entry.name === input.name) !== -1) {
    throw (HTTPError(400));
  }

  if (input.type === "recipe") {
    const temp: string[] = (input as recipe).requiredItems.map((item) => item.name);
    if (new Set(temp).size !== temp.length) {
      throw (HTTPError(400));
    }

    cookbook.push(input as recipe);
  } else {
    if ((input as ingredient).cookTime < 0) {
      throw (HTTPError(400));
    }

    cookbook.push(input as ingredient);
  }
}

// [TASK 3] ====================================================================
// Endpoint that returns a summary of a recipe that corresponds to a query name
app.get("/summary", (req:Request, res:Request) => {
  const input : string = req.query.name;

  res.status(200).send(get_summary(input));
});

// Takes in a recipe namee and returns the summary of the recipe
const get_summary = (input: string) : summary => {

  let recipe : summary = cookbook.find((entry) => entry.name === input);
  if (!recipe || recipe.type === "ingredient") {
    throw (HTTPError(400));
  }

  let summary: summary = {...recipe};

  summary.requiredItems = recursiveSummary(summary.requiredItems);
  summary.cookTime = 0;

  // sums up the cookTime of the recipe
  for (let i = 0; i < summary.requiredItems.length; i++) {
    const item = summary.requiredItems[i];
    summary.cookTime += cookbook.find((entry) => entry.name === item.name).cookTime * item.quantity
  }

  return summary;
}

// recursively gets the list of all required items for the recipe
const recursiveSummary = (requiredItems: requiredItem[]) : requiredItem[] => {
  let result : requiredItem[] = [...requiredItems];

  for (const item of requiredItems) {
    const recipeReference: cookbookEntry = 
      cookbook.find((entry) => entry.name === item.name);

    if (!recipeReference) {
      throw(HTTPError(400));
    } else if (recipeReference.type === "ingredient") continue;
    
    const temp : requiredItem[] = recursiveSummary((recipeReference as recipe).requiredItems);
    mergeRequiredItems(result, temp, item.quantity);

    result.splice(result.findIndex((temp) => temp.name === item.name), 1);
  }

  return result;
}

// merges the required items of 2 different recipes.
const mergeRequiredItems = (arr1: requiredItem[], arr2: requiredItem[], quantity: number)=> {
  for (const entry of arr2) {
    if (arr1.find((it) => it.name === entry.name)) {
      arr1.find((it) => it.name === entry.name).quantity += entry.quantity * quantity;
    } else {
      arr1.push({...entry, quantity: entry.quantity * quantity})
    }
  }
}

// =============================================================================
// ==== DO NOT TOUCH ===========================================================
// =============================================================================
const port = 8080;
app.listen(port, () => {
  console.log(`Running on: http://127.0.0.1:8080`);
});

File: backend/py_template/devdonalds.py
from dataclasses import dataclass
from typing import List, Dict, Union
from flask import Flask, request, jsonify
import re

# ==== Type Definitions, feel free to add or modify ===========================
@dataclass
class CookbookEntry:
	name: str
	type: str

@dataclass
class RequiredItem():
	name: str
	quantity: int

@dataclass
class Recipe(CookbookEntry):
	required_items: List[RequiredItem]

@dataclass
class Ingredient(CookbookEntry):
	cook_time: int

class Summary(CookbookEntry):
	required_items: List[RequiredItem]
	cook_time: int

# =============================================================================
# ==== HTTP Endpoint Stubs ====================================================
# =============================================================================
app = Flask(__name__)

# Store your recipes here!
cookbook = []

# Task 1 helper (don't touch)
@app.route("/parse", methods=['POST'])
def parse():
	data = request.get_json()
	recipe_name = data.get('input', '')
	parsed_name = parse_handwriting(recipe_name)
	if parsed_name is None:
		return 'Invalid recipe name', 400
	return jsonify({'msg': parsed_name}), 200

# [TASK 1] ====================================================================
# Takes in a recipeName and returns it in a form that 
def parse_handwriting(recipeName: str) -> Union[str | None]:
	recipeName = re.sub('[-_]', ' ', recipeName)
	recipeName = recipeName.lower()
	recipeName = re.sub('[^a-z ]', "", recipeName)
	recipeName = re.sub(' +', " ", recipeName)
	recipeName = recipeName.title()
	recipeName = recipeName.strip()
	
	if len(recipeName) == 0 :
		return None
	 
	return recipeName


# [TASK 2] ====================================================================
# Endpoint that adds a CookbookEntry to your magical cookbook
@app.route('/entry', methods=['POST'])
def entry():
	data = request.get_json()
	# recipe_name = data.get('input', '')
	entry = createCookbookEntry(data)
	if entry is not None:
		return entry, 400	 
	return jsonify({}), 200

# Takes in a input and either returns an error or adds the entry to input
def createCookbookEntry(input: CookbookEntry):
	if not (input['type'] == 'recipe' or input['type'] == 'ingredient'):	
		return "Input type is invalid"
	elif any(entry for entry in cookbook if entry['name'] == input['name']):
		return "Entry already exists"

	if input['type'] == 'recipe':
		input['required_items'] = input['requiredItems']
		del input['requiredItems']

		temp = [entry['name'] for entry in input['required_items']]
		if len(temp) != len(set(temp)):
			return 'required Items can only have one element per name'
	elif input['type'] == 'ingredient':
		input['cook_time'] = input['cookTime']
		del input['cookTime']

		if input['cook_time'] < 0:
			return 'cook_Time cannot be negative'

	cookbook.append(input)
	return None

# [TASK 3] ====================================================================
# Endpoint that returns a summary of a recipe that corresponds to a query name
@app.route('/summary', methods=['GET'])
def summary():
	name = request.args.get('name')

	summary = create_summary(name)
	if type(summary) is str:
		return summary, 400	 
	return jsonify(summary), 200

# Takes in a recipeName and returns an a summary of the recipe
def create_summary(name: str) -> Summary:
	recipe: CookbookEntry = next(iter([entry for entry in cookbook if entry['name'] == name]), None)

	if recipe == None:
		return "recipe does not exist"
	elif recipe['type'] == 'ingredient':
		return "searched name is not a recipe name"
	
	summary: Summary = recipe
	temp: Union[str | Summary]  = recursive_summary(recipe['required_items'])

	if type(temp) == str:
		return temp

	summary['required_items'] = temp
	summary['cook_time'] = 0
	
	# sums up cookTime for every item in required_items
	for item in summary['required_items']: 
		ingredient: Ingredient = next(iter([entry for entry in cookbook if entry['name'] == name]))
		recipe['cook_time'] += item['quantity'] * ingredient['cook_time']

	return summary

def recursive_summary(required_items: List[RequiredItem]) -> Union[str | Summary]:
	result: List[RequiredItem] = required_items.copy()

	for item in result:
		cookbook_entry: CookbookEntry = next(iter([x for x in cookbook if x['name'] == item['name']]), None)
		
		if cookbook_entry == None:
			return "recipe contains recipe or ingredients that don't exist"
		elif cookbook_entry['type'] == 'ingredient':
			continue
		
		# recursively finds and adds ingredients from recipe
		temp: Union[str | Summary] = recursive_summary(cookbook_entry['required_items'])
		
		if type(temp) == str:
			return temp
		
		merge_required_item(result, temp, item['quantity'])
		
		# remove recipe from result array
		result = [x for x in result if x['name'] != item['name']]

	return result

# merges the required items of 2 different recipes 
def merge_required_item(arr1: List[RequiredItem], arr2: List[RequiredItem], quantity: int):
	for entry in arr2:
		item = None

		for x in arr1:
			if x['name'] == entry['name']:
				item = x
				break
		
		if item == None:
			item = entry.copy()
			item['quantity'] = entry['quantity'] * quantity
			arr1.append(item)
		else:
			item['quantity'] += entry['quantity'] * quantity


# =============================================================================
# ==== DO NOT TOUCH ===========================================================
# =============================================================================

if __name__ == '__main__':
	app.run(debug=True, port=8080)
